src/lm_speed.cpp

```diff
@@ -0,0 +1,502 @@
+      Eigen::VectorXd effects(PQR.householderQ().adjoint() * y);
+      beta_out.head(r) = R_inv * effects.head(r);
+      // Rcout << beta_out << std::endl << std::endl;
+      beta_out = Pmat * beta_out;
+
+      // Get all column indices
+      Eigen::ArrayXi Pmat_indices = Pmat.indices();
+      // Get the order for the columns you are keeping
+      Eigen::ArrayXi Pmat_keep = Pmat_indices.head(r);
+      // Get the indices for columns you are discarding
+      Eigen::ArrayXi Pmat_toss = Pmat_indices.tail(p - r);
+
+      // this code takes the indices you are keeping, and, while preserving order, keeps them in the range [0, r-1]
+      // For each one, see how many dropped indices are smaller, and subtract that difference
+      // Ex: p = 4, r = 2
+      // Pmat_indices = {3, 1, 0, 2}
+      // Pmat_keep = {3, 1}
+      // Pmat_toss = {0, 2}
+      // Now we go through each keeper, count how many in toss are smaller, and then modify accordingly
+      // 3 - 2 and 1 - 1
+      // Pmat_keep = {1, 0}
+      for(Eigen::Index i=0; i<r; ++i)
+      {
+        Pmat_keep(i) = Pmat_keep(i) - (Pmat_toss < Pmat_keep(i)).count();
+      }
+      // Rcout << Pmat_indices << std::endl << std::endl;
+      // Rcout << Pmat_keep << std::endl << std::endl;
+      //
+      // Rcout << Pmat * Eigen::MatrixXd::Identity(p, p) << std::endl << std::endl;
+
+      Eigen::PermutationMatrix<Eigen::Dynamic, Eigen::Dynamic> P = Eigen::PermutationWrapper<Eigen::ArrayXi>(Pmat_keep);
+      // Rcout << P * Eigen::MatrixXd::Identity(r, r) << std::endl << std::endl;
+
+      R_inv = P * R_inv * P;
+      XtX_inv = R_inv * R_inv.transpose();
+      // Rcout << XtX_inv << std::endl;
+    }
+  }
+
+  Eigen::MatrixXd Xoriginal(n, r);
+  Eigen::MatrixXd Xoriginalfull(n, p);
+  Eigen::MatrixXd X(n, r);
+  Eigen::VectorXd beta_hat(r);
+  Eigen::ArrayXd weights;
+  bool weighted = false;
+  if (Xunweighted.isNotNull()) {
+    weights = Rcpp::as<Eigen::Map<Eigen::ArrayXd> >(weight);
+    Xoriginalfull = Rcpp::as<Eigen::Map<Eigen::MatrixXd> >(Xunweighted);
+    weighted = true;
+  }
+  // Reshape objects if rank-deficient
+  if (r < p) {
+    // Rcout << "Rank-deficient!" << std::endl;
+    // Rcout << "beta_out:" << beta_out << std::endl;
+
+    unsigned j = 0;
+    for (unsigned i = 0; i < p; i++) {
+      // Rcout << "isnan: " << std::isnan(beta_out(i)) << std::endl;
+      if (!std::isnan(beta_out(i))) {
+        // Rcout << "not NA: " << i << std::endl;
+        beta_hat(j) = beta_out(i);
+        X.col(j) = Xfull.col(i);
+        if (weighted) {
+          Xoriginal.col(j) = Xoriginalfull.col(i);
+        } else {
+          Xoriginal.col(j) = Xfull.col(i);
+        }
+        j++;
+      }
+    }
+  } else {
+    beta_hat = beta_out;
+    X = Xfull;
+    if (weighted) {
+      Xoriginal = Xoriginalfull;
+    } else {
+      Xoriginal = Xfull;
+    }
+  }
+
```

Comment:
Here is where I construct all of my matrices for the work below. Note that I am just dropping the columns of X that have been dropped by the QR, and removing the NAs from beta.